odoo/osv/expression.py

```diff
@@ -1083,6 +1083,14 @@ def push_result(sql):
                             "((%s) %s (%s))",
                             unaccent(sql_left), sql_operator, unaccent(sql_right),
                         ))
+            elif field.type in ('datetime', 'date') and len(path) == 2:
+                if path[1] not in READ_GROUP_NUMBER_GRANULARITY:
+                    raise ValueError(f'Error when processing the field {field!r}, the granularity {path[1]} is not supported. Only "week_number", "month_number" and "quarter_number" are supported')
+                sql_field = model._field_to_sql(alias, field.name, self.query)
+                if model._context.get('tz') in pytz.all_timezones_set:
+                    sql_field = SQL("timezone(%s, timezone('UTC', %s))", model._context['tz'], sql_field)
+                sql = SQL('date_part(%s, %s) %s %s', READ_GROUP_NUMBER_GRANULARITY[path[1]], sql_field, SQL_OPERATORS[operator], int(right))
```

Comment:
Why forcing `int(right)`?  This isn't necessary and prevents us from using an `SQL` object as the right-hand side value.